storage: declare fewer spans during GC

tbg opened the pull request:
If a GCRequest does not modify the GC threshold timestamps, it does
not need to declare their keys. This is useful because it allows the
GC queue to send a very small request with the threshold adjustments
first, and then a larger request that does not need to declare the
threshold keys and does less harm if it's slow.

We should additionally limit the size of each "big" `GCRequest`, but
that fits well in a follow-up PR.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16210)
<!-- Reviewable:end -->

tamird commented:
What are the atomicity implications here? In other words, what if the timestamps get bumped but the subsequent "real" `GCRequest` fails?

---

Reviewed 4 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/keys/keys.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPgcna7nyre_5yNAkm:-KlPgcna7nyre_5yNAkn:b85mtgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/keys/keys.go#L203)):*
> this timestamp will not be served.

Can you be more specific? What will be the error? is it even appropriate to mention this on this function, which only knows of the key and not its value?

---

*[pkg/storage/gc_queue.go, line 563 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPmFPcnfwfDi6x7Iny:-KlPmFPcnfwfDi6x7Inz:blxsoqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/gc_queue.go#L563)):*
> ```Go
> 
> 	batches := func() []roachpb.GCRequest {
> 		var template roachpb.GCRequest
> ```

why not a composite literal? I don't think it matters, just wondering.

---

*[pkg/storage/gc_queue.go, line 582 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPmf36VPQbZON6DscT:-KlPmf377U3HB_DRDHpN:bg66bh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/gc_queue.go#L582)):*
> ```Go
> 		// Technically not needed since we're talking directly to the Range.
> 		ba.RangeID = desc.RangeID
> 		ba.Timestamp = now
> ```

perhaps a newline above this line to clarify the comment.

---

*[pkg/storage/gc_queue.go, line 589 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPmlFFruBHAU960eOL:-KlPmlFFruBHAU960eOM:bs5jag8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/gc_queue.go#L589)):*
> ```Go
> 		// in a batch which might end up spanning Ranges by the time it executes.
> 		ba.Add(&gcArgs)
> 		log.Eventf(ctx, "sending batch %d of %d", i, len(batches))
> ```

is it worth having an event on either side, so we can see how long each batch takes?

---

*[pkg/storage/replica_command.go, line 1433 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPn3mbVX5m8Jssywbn:-KlPn3mbVX5m8Jssywbo:blzxdew) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/replica_command.go#L1433)):*
> ```Go
> 	var newThreshold hlc.Timestamp
> 	if args.Threshold != (hlc.Timestamp{}) {
> ```

nit: why this empty line?

---

*[pkg/storage/replica_command.go, line 1452 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPnBhPEp5Y4NRBfarD:-KlPnBhPEp5Y4NRBfarE:b-vd2q60) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/replica_command.go#L1452)):*
> ```Go
> 	}
> 
> 	stateLoader := makeReplicaStateLoader(cArgs.EvalCtx.RangeID())
> ```

moving this above `var pd EvalResult` is just diff noise, I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-:-KlPmt2tacb8kvQ0EHxa:b4z3gum)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
No implications other than theoretically being able to serve lower timestamps. Reversing the order of the batches would be incorrect, of course.

---

Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/keys/keys.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPgcna7nyre_5yNAkm:-KlPopGKBHX79biHRctV:b-slraay) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/keys/keys.go#L203)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> this timestamp will not be served.

Can you be more specific? What will be the error? is it even appropriate to mention this on this function, which only knows of the key and not its value?
</blockquote></details>

The method is poorly named, and I've had to look up the details a couple of times (hence the TODO below). What would you like to read here? Reads below GCThreshold isn't possible because we've deleted the data.

---

*[pkg/storage/gc_queue.go, line 563 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPmFPcnfwfDi6x7Iny:-KlPp3R2GH_lGPznj7oE:b-1ohemm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/gc_queue.go#L563)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why not a composite literal? I don't think it matters, just wondering.
</blockquote></details>

`Key` and `EndKey` are in embedded structs, so making it composite causes eye cancer.

---

*[pkg/storage/gc_queue.go, line 582 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPmf36VPQbZON6DscT:-KlPpFO4r6LOTFTchwot:b-gqsyfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/gc_queue.go#L582)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

perhaps a newline above this line to clarify the comment.
</blockquote></details>

The comment applies to both. You don't need to set a timestamp for nontransactional batches (at least not last I checked).

---

*[pkg/storage/gc_queue.go, line 589 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPmlFFruBHAU960eOL:-KlPpNWxnYmyDmau3rdV:bvtgwqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/gc_queue.go#L589)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is it worth having an event on either side, so we can see how long each batch takes?
</blockquote></details>

You do, either from the loop or the final one below. In my experience, less is often more with traces as they get more annoying to parse as they get noisier.

---

*[pkg/storage/replica_command.go, line 1433 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPn3mbVX5m8Jssywbn:-KlPpXsFAH19XMplCDAR:b-ris6r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/replica_command.go#L1433)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: why this empty line?
</blockquote></details>

Accident. Removed.

---

*[pkg/storage/replica_command.go, line 1452 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPnBhPEp5Y4NRBfarD:-KlPpwnVLqW5sErPh_eY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/replica_command.go#L1452)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

moving this above `var pd EvalResult` is just diff noise, I think.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-:-KlPohjjXJdn6TGgKIvk:b-sx8u7g)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
And what protects against the batches being processed out of order? We're trying to handle some reorderings by tracking the maximum timestamps. (This, incidentally, is what makes it possible to make this change - if we didn't ratchet the last GC timestamp forward then this change would result in the timestamp being set to zero during an upgrade). I think we should refuse to process any GCKeys whose timestamp is ahead of the current GC threshold. 

---

Reviewed 5 of 5 files at r1.
Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_command.go, line 1373 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPoOVYHfm0nYqTnWLk:-KlPoOVZ-3g3iPCOyOM6:b5xsxsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/replica_command.go#L1373)):*
> ```Go
> func declareKeysGC(
> 	desc roachpb.RangeDescriptor, header roachpb.Header, req roachpb.Request, spans *SpanSet,
> ) {
> ```

Comment that we deliberately do not call DefaultDeclareKeys, which would cause any GC request to block the entire range. This is non-obvious: I wrote all this code and didn't even realize that we weren't automatically blocking the whole range because of the header span. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-:-KlPq6CvJ2zoSegAgurY:bojoabv)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell `GCThreshold` and `TxnSpanGCThreshold` are non-nullable fields, so every Raft command sends them as zero. We don't only ratchet forward in `evalGC` but also in the state update, btw. Can you elaborate on your concern here? We're GC'ing keys only after certifiably having moved the thresholds forward.

BTW, I got curious about what splits do with these thresholds and it seems to me that they don't copy it to the new RHS of the split, which amounts to a bug (and one very similar to what has cost us a lot of time already with the leases). I'll file that unless you find I'm wrong.

---

Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica_command.go, line 1373 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPoOVYHfm0nYqTnWLk:-KlPubx6x0JoubaMGiC0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/storage/replica_command.go#L1373)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that we deliberately do not call DefaultDeclareKeys, which would cause any GC request to block the entire range. This is non-obvious: I wrote all this code and didn't even realize that we weren't automatically blocking the whole range because of the header span. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-:-KlPsFgoULX2Hn_WVzgp:b-iktmiu)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/keys/keys.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPgcna7nyre_5yNAkm:-KlPuryzEWTIBeKYJGvP:b-60358q) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/keys/keys.go#L203)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The method is poorly named, and I've had to look up the details a couple of times (hence the TODO below). What would you like to read here? Reads below GCThreshold isn't possible because we've deleted the data.
</blockquote></details>

"Will not be served" is ambiguous, though. Particularly after this PR, where the two parts of GC are processed separately - would a read that interleaves the two be served? if not, what would be returned?

---

*[pkg/storage/replica_test.go, line 7265 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPvMkcfPvlppa5GMCE:-KlPvMkdhW3F_vSnV3qp:b-cxrb3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cffe8d96902a82415e0fd9a6ba905205478568b/pkg/storage/replica_test.go#L7265)):*
> ```Go
> 					},
> 				}, &resp); err != nil {
> 					t.Fatal(errors.Wrapf(err, "at (%s,%s)", keyThresh, txnThresh))
> ```

doesn't matter much, but this can easily just be `Fatalf` and get a shorter line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Yeah, I guess with the ratcheting there is no reordering scenario that could cause problems.

It sounds like a bug that we don't copy these on splits. We haven't done much to test these thresholds; we've been assuming that because the default GC threshold is a day there's not much risk of people pushing the limits with these timestamps. We should probably do more to test that historical queries don't silently return incomplete data, since we've gone to all this trouble to track these timestamps.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-:-KlPvKzYDgqC6BImP5Uy:b-p5o2rc)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR! Filed https://github.com/cockroachdb/cockroach/issues/16216 about the missing split initialization.

---

Review status: 4 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/keys/keys.go, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPgcna7nyre_5yNAkm:-KlPx8_gsv-lXXfUAjjL:bn0lh08) ([raw file](https://github.com/cockroachdb/cockroach/blob/2cf2cda4f9ef6291c79ec61dc8e56155d434c80a/pkg/keys/keys.go#L203)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"Will not be served" is ambiguous, though. Particularly after this PR, where the two parts of GC are processed separately - would a read that interleaves the two be served? if not, what would be returned?
</blockquote></details>

Only the first part of GC updates this timestamp. Between the two requests, it would be legal to serve reads at the old threshold, but this is not done. The comment is here primarily to assure me until the key is renamed (which I don't want to do during cherrypicking season). Please entertain me.

---

*[pkg/storage/replica_test.go, line 7265 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPvMkcfPvlppa5GMCE:-KlPway4e83HOsUuWwXF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cffe8d96902a82415e0fd9a6ba905205478568b/pkg/storage/replica_test.go#L7265)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

doesn't matter much, but this can easily just be `Fatalf` and get a shorter line
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-:-KlPwc-5jKI-GtPmMB9d:b-6ue6bn)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_test.go, line 7265 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPyKfED0bSb5IutSxU:-KlPyKfFRdh2tR7LCMQL:b-hpwuqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cffe8d96902a82415e0fd9a6ba905205478568b/pkg/storage/replica_test.go#L7265)):*
> ```Go
> 					},
> 				}, &resp); err != nil {
> 					t.Fatal(errors.Wrapf(err, "at (%s,%s)", keyThresh, txnThresh))
> ```

you lost the error here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: declare fewer spans during GC

If a GCRequest does not modify the GC threshold timestamps, it does
not need to declare their keys. This is useful because it allows the
GC queue to send a very small request with the threshold adjustments
first, and then a larger request that does not need to declare the
threshold keys and does less harm if it's slow.

We should additionally limit the size of each "big" `GCRequest`, but
that fits well in a follow-up PR.

tschottdorf commented:
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica_test.go, line 7265 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210#-KlPyKfED0bSb5IutSxU:-KlPz7kQvX79Y_u4SePZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5cffe8d96902a82415e0fd9a6ba905205478568b/pkg/storage/replica_test.go#L7265)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you lost the error here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16210)*
<!-- Sent from Reviewable.io -->
